services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala

```diff
@@ -0,0 +1,183 @@
+  val metricServiceSettings = MetricServiceSettings.newBuilder.setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build
+  final val metricServiceClient = MetricServiceClient.create(metricServiceSettings)
+
+  // Prepares the monitored resource descriptor
+  lazy val resourceLabels = Map[String, String](("project_id", stackdriverConfig.googleProject))
+  lazy val monitoredResource = MonitoredResource.newBuilder.setType("global").putAllLabels(resourceLabels.asJava).build
+
+  // Give the actor time to warm up, then start sending the metrics to Stackdriver at an interval
+  context.system.scheduler.schedule(InitialDelay, stackdriverConfig.flushRate, self, SendStackdriverMetricCommand)
+
+
+  override def receive = {
+    case SendStackdriverMetricCommand => sendMetricData()
+    case InstrumentationServiceMessage(cromwellMetric) => cromwellMetric match {
+      case CromwellTiming(bucket, value, _) => updateMetricMap(bucket, value.toMillis.toDouble, StackdriverGauge)
+      case CromwellGauge(bucket, value) => updateMetricMap(bucket, value.toDouble, StackdriverGauge)
+      case CromwellCount(bucket, value, _) => updateMetricMap(bucket, value.toDouble, StackdriverCumulative)
+      case CromwellIncrement(bucket) => updateMetricMap(bucket, metricValue = 1D, metricKind = StackdriverCumulative)
+    }
+    case ShutdownCommand =>
+      // flush out metrics (if any) before shut down
+      sendMetricData()
+      metricServiceClient.close()
+      context stop self
+  }
+
+
+  private def labelFromConfig(op: StackdriverConfig => Option[String], key: String): List[(String, String)] = {
+    op(stackdriverConfig).fold(List.empty[(String, String)])(v => List((key.replace("-", "_"), v)))
+  }
+
+
+  private def generateMetricLabels(): Map[String, String] = {
+    val labelsList = labelFromConfig(_.cromwellInstanceIdentifier, CromwellInstanceIdentifier) ++
+      labelFromConfig(_.cromwellInstanceRole, CromwellInstanceRole) ++
+      labelFromConfig(_.cromwellPerfTestCase, CromwellPerfTest)
+
+    labelsList.toMap
+  }
+
+
+  private def updateMetricMap(bucket: CromwellBucket, metricValue: Double, metricKind: StackdriverMetricKind): Unit = {
+    val metricObj = StackdriverMetric(bucket.toStackdriverString, metricKind)
+
+    if (metricsMap.contains(metricObj)) {
+      val valueList = metricsMap(metricObj)
+      metricsMap += metricObj ->  valueList.::(metricValue)
+    }
+    else metricsMap += metricObj -> List(metricValue)
+  }
+
+
+  private def sendMetricData(): Unit = {
+    metricsMap.foreach { case (key, value: List[Double]) =>
+      val dataPointListSum = value.sum
+      val dataPoint = key.kind match {
+        case StackdriverGauge => dataPointListSum / value.length
+        case StackdriverCumulative => dataPointListSum
+      }
+      writeMetrics(key, dataPoint)
+    }
+
+    metricsMap = Map.empty[StackdriverMetric, List[Double]]
+  }
+
+
+  private def timeInterval(metricKind: StackdriverMetricKind): TimeInterval = {
```

Comment:
TOL: For these three helper methods (`timeInterval`, `createTimeSeries` and `createMetricBuilder`) which are only ever used in `writeMetrics`, my preference is to move them inside that method so that it's obvious that they only exist to help that method do its work.